src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,571 @@
+        position as possible.
+        @dev msg.sender must be authorized.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function cage(bytes32 ilk) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-cage-during-shutdown");
+
+        ilks[ilk].tic = block.timestamp;
+        emit Cage(ilk);
+    }
+
+    /**
+        @notice Write off the debt for a caged pool.
+        This must occur while vat is live and after tau number of seconds has
+        passed since the pool was caged.
+        @dev This will send the pool's debt to the vow as sin and convert its
+        collateral to gems.  There is a situation where another user has paid
+        back some of the Pool's debt where ink != art in this case we rebalance
+        so that vat.gems(pool) will represent the amount of debt sent to the vow.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function cull(bytes32 ilk) external {
+        require(vat.live() == 1, "DssDirectDepositHub/no-cull-during-shutdown");
+
+        uint256 tic = ilks[ilk].tic;
+        require(tic > 0, "DssDirectDepositHub/pool-live");
+
+        uint256 tau = ilks[ilk].tau;
+        require(_add(tic, tau) <= block.timestamp || wards[msg.sender] == 1, "DssDirectDepositHub/unauthorized-cull");
+
+        uint256 culled = ilks[ilk].culled;
+        require(culled == 0, "DssDirectDepositHub/already-culled");
+
+        D3MPoolLike pool = ilks[ilk].pool;
+
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(pool));
+        require(ink <= 2 ** 255, "DssDirectDepositHub/overflow");
+        require(art <= 2 ** 255, "DssDirectDepositHub/overflow");
+        vat.grab(ilk, address(pool), address(pool), vow, -int256(ink), -int256(art));
+
+        if (ink > art) {
+            // We have more collateral than debt, so need to rebalance.
+            // After cull the gems we grab above represent the debt to
+            // unwind.
+            vat.slip(ilk, address(pool), -int256(ink - art));
+        }
+
+        ilks[ilk].culled = 1;
+        emit Cull(ilk);
+    }
+
+    /**
+        @notice Rollback Write-off (cull) if General Shutdown happened.
+        This function is required to have the collateral back in the vault so it
+        can be taken by End module and eventually be shared to DAI holders (as
+        any other collateral) or maybe even unwinded.
+        @dev This pulls gems from the pool and reopens the urn with the gem
+        amount of ink/art.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function uncull(bytes32 ilk) external {
+        D3MPoolLike pool = ilks[ilk].pool;
+
+        require(ilks[ilk].culled == 1, "DssDirectDepositHub/not-prev-culled");
+        require(vat.live() == 0, "DssDirectDepositHub/no-uncull-normal-operation");
+
+        address vow_ = vow;
+        uint256 wad = vat.gem(ilk, address(pool));
+        require(wad < 2 ** 255, "DssDirectDepositHub/overflow");
```

Comment:
Discussed in review we should use `require(int256(wad) >= 0)` throughout, and make it a team best-practice generally.